hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMBucketCreateRequest.java

```diff
@@ -0,0 +1,203 @@
+    .BucketEncryptionInfoProto;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .CreateBucketRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .CreateBucketResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .BucketInfo;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .OMRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .OMResponse;
+import org.apache.hadoop.ozone.protocolPB.OMPBHelper;
+import org.apache.hadoop.ozone.om.ratis.utils.OzoneManagerRatisUtils;
+import org.apache.hadoop.util.Time;
+import org.apache.hadoop.utils.db.cache.CacheKey;
+import org.apache.hadoop.utils.db.cache.CacheValue;
+
+import static org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .CryptoProtocolVersionProto.ENCRYPTION_ZONES;
+
+/**
+ * Handles CreateBucket Request.
+ */
+public class OMBucketCreateRequest extends OMClientRequest {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMBucketCreateRequest.class);
+
+  public OMBucketCreateRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+
+    // Get original request.
+    CreateBucketRequest createBucketRequest =
+        getOmRequest().getCreateBucketRequest();
+    BucketInfo bucketInfo = createBucketRequest.getBucketInfo();
+
+    // Get KMS provider.
+    KeyProviderCryptoExtension kmsProvider =
+        ozoneManager.getKmsProvider();
+
+    // Create new Bucket request with new bucket info.
+    CreateBucketRequest.Builder newCreateBucketRequest =
+        createBucketRequest.toBuilder();
+
+    BucketInfo.Builder newBucketInfo = bucketInfo.toBuilder();
+
+    newCreateBucketRequest.setBucketInfo(
+        newBucketInfo.setCreationTime(Time.now()));
+
+    if (bucketInfo.hasBeinfo()) {
+      newBucketInfo.setBeinfo(getBeinfo(kmsProvider, bucketInfo));
+    }
+
+    newCreateBucketRequest.setBucketInfo(newBucketInfo.build());
+    return getOmRequest().toBuilder().setCreateBucketRequest(
+        newCreateBucketRequest.build()).build();
+  }
+
+  @Override
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long transactionLogIndex) {
+    OMMetrics omMetrics = ozoneManager.getMetrics();
+    omMetrics.incNumBucketCreates();
+
+    OMMetadataManager metadataManager = ozoneManager.getMetadataManager();
+
+    BucketInfo bucketInfo = getBucketInfoFromRequest();
+
+    String volumeName = bucketInfo.getVolumeName();
+    String bucketName = bucketInfo.getBucketName();
+
+    OMResponse.Builder omResponse = OMResponse.newBuilder().setCmdType(
+        OzoneManagerProtocolProtos.Type.CreateBucket).setStatus(
+        OzoneManagerProtocolProtos.Status.OK);
+    OmBucketInfo omBucketInfo = null;
+
+
+    metadataManager.getLock().acquireVolumeLock(volumeName);
```

Comment:
Why do we need a volume lock here?

Reply:
Acquired volume lock here, so that deleteVolume cannot happen while creating bucket.

Take a scenario: volume created, then delete Volume and Create bucket in that volume came in parallel. If we don't acquire volume lock, there might be a chance to create a bucket in a deletd volume.